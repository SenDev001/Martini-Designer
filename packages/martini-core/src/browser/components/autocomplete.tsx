import * as React from "react";
import styled from "styled-components";
import { ElementExt } from '@phosphor/domutils';

export interface Suggestion {
    label: string;
    iconClass?: string;
}

export interface AutocompleteProps {
    value: string;
    suggestions: Suggestion[];
    onInputChange: (input: string) => void;
    onInputBlur?: (e: React.FocusEvent<HTMLInputElement>, input: string) => void;
    onSuggestionSelect?: (input: string) => void;
    onInputKeyDown?: (e: React.KeyboardEvent<HTMLInputElement>) => void;
    onInputKeyUp?: (e: React.KeyboardEvent<HTMLInputElement>) => void;
    inputStyle?: React.CSSProperties;
    autoFocus?: boolean;
}

const UlStyles = styled.ul`
    color: var(--theia-input-foreground);
    background-color: var(--theia-input-background);
    border: 1px solid var(--theia-inputOption-activeBorder);
    border-top: 0;
    position: absolute;
    top: 25px;
    left: 0px;
    width: calc(100% - 2px);
    list-style: none;
    margin-top: 0;
    max-height: 143px;
    overflow-y: auto;
    padding-left: 0;
    z-index: 20;

    li {
        padding: 4px;
    }

    .suggestion-active, li:hover {
        background-color: var(--theia-list-activeSelectionBackground);
        color: var(--theia-input-foreground);
        cursor: pointer;
    }
`;

export const Autocomplete: React.FC<AutocompleteProps> = ({
    value,
    suggestions,
    onInputChange,
    onInputBlur,
    onSuggestionSelect,
    onInputKeyDown,
    onInputKeyUp,
    inputStyle,
    autoFocus
}) => {
    const [activeSuggestion, setActiveSuggestion] = React.useState(0);
    const [filteredSuggestions, setFilteredSuggestions] = React.useState([] as Suggestion[]);
    const [showSuggestions, setShowSuggestions] = React.useState(false);
    const mounted = React.useRef(true);
    React.useEffect(() => {
        if (autoFocus && inputRef.current)
            inputRef.current.setSelectionRange(0, value.length);
        return () => { mounted.current = false; };
    }, []);
    const inputRef = React.useRef<HTMLInputElement>(null);
    const ulRef = React.useRef<HTMLUListElement>(null);

    const filter = (input: string) => input.length === 0 ? suggestions : suggestions.filter(
        suggestion =>
            suggestion.label.toLowerCase().includes(input)
    );

    const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        const input = e.target.value;
        setActiveSuggestion(0);
        setFilteredSuggestions(filter(input.toLowerCase()));
        setShowSuggestions(true);
        onInputChange(input);
    };

    const onClick = (suggestion: Suggestion) => {
        setActiveSuggestion(0);
        setFilteredSuggestions([]);
        setShowSuggestions(false);
        onInputChange(suggestion.label);
        if (onSuggestionSelect)
            onSuggestionSelect(suggestion.label);
        inputRef.current?.focus();
    };

    const onKeyUp = (e: React.KeyboardEvent<HTMLInputElement>) => {
        if (e.key === "Enter" && showSuggestions && filteredSuggestions.length !== 0) {
            const label = filteredSuggestions[activeSuggestion].label;
            setActiveSuggestion(0);
            setShowSuggestions(false);
            onInputChange(label);
            if (onSuggestionSelect)
                onSuggestionSelect(label);
            return;
        }
        else if (e.key === "Escape" && showSuggestions) {
            setShowSuggestions(false);
            return;
        }
        else if (e.key === "ArrowUp") {
            e.preventDefault();
            let next = activeSuggestion - 1;

            if (next < 0)
                next = filteredSuggestions.length - 1;

            setActiveSuggestion(next);
            if (ulRef.current)
                ElementExt.scrollIntoViewIfNeeded(ulRef.current, ulRef.current.childNodes[next] as HTMLLIElement);
            return;
        }
        else if (e.key === "ArrowDown") {
            e.preventDefault();
            let next = activeSuggestion + 1;

            if (next >= filteredSuggestions.length)
                next = 0;

            setActiveSuggestion(next);
            if (ulRef.current)
                ElementExt.scrollIntoViewIfNeeded(ulRef.current, ulRef.current.childNodes[next] as HTMLLIElement);
            return;
        }
        else if (!showSuggestions && e.key === " " && e.ctrlKey) {
            setShowSuggestions(true);
            return;
        }

        if (onInputKeyUp)
            onInputKeyUp(e);
    };

    const onBlur = (e: React.FocusEvent<HTMLInputElement>) => {
        // when e.relatedTarget === null it means it probably click one of the suggestions
        if (e.relatedTarget !== null && onInputBlur) {
            onInputBlur(e, value);
            setShowSuggestions(false);
        }
    };

    let suggestionsListComponent;

    if (showSuggestions && filteredSuggestions.length) {
        suggestionsListComponent = (
            <UlStyles ref={ulRef}>
                {filteredSuggestions.map((suggestion, index) => {
                    return (
                        <li
                            className={index === activeSuggestion ? "suggestion-active" : undefined}
                            key={suggestion.label}
                            onClick={() => onClick(suggestion)}
                        >
                            {suggestion.label}
                        </li>
                    );
                })}
            </UlStyles>
        );
    }

    return <>
        <input
            type="text"
            autoFocus={autoFocus}
            ref={inputRef}
            onChange={onChange}
            onBlur={onBlur}
            value={value}
            onKeyDown={onInputKeyDown}
            onKeyUp={onKeyUp}
            className="theia-input"
            style={inputStyle}
        />
        {suggestionsListComponent}
    </>;
};
