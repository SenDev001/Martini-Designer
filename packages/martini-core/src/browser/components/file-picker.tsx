import * as React from "react";

export interface FilePickerProps {
    extensions?: string;
    placeholder?: string;
    value?: File | string;
    onChange?: (fileOrUrl: File | string) => void;
}

export const FilePicker: React.FC<FilePickerProps> = ({extensions, placeholder, value, onChange, children}) => {
    const [file, setFile] = React.useState(!value ? "" : value instanceof File ? value.name : value);
    const fileInput = React.useRef<HTMLInputElement>(null);
    const [invalidDrop, setInvalidDrop] = React.useState(false);

    const isValidFile = (file: File) => {
        if (!extensions)
            return true;
        return extensions.split(/,/).some(ext => file.name.endsWith(ext));
    };
    const handleFiles = (files: FileList) => {
        if (files && files.length !== 0) {
            const first = files.item(0);
            if (first && isValidFile(first)) {
                setFile(first.name);
                if (onChange)
                    onChange(first);
                return true;
            }
        }
    };
    const handleFileInputChanged = (e: React.ChangeEvent<HTMLInputElement>) => {
        const files = e.target.files;
        if (files)
            handleFiles(files);
    };
    const handleTextInputChanged = (e: React.ChangeEvent<HTMLInputElement>) => {
        setFile(e.target.value);
        onChange!(e.target.value);
    };

    const handleDrop = (e: React.DragEvent) => {
        if (!handleFiles(e.dataTransfer.files)) {
            e.dataTransfer.dropEffect = "none";
            e.dataTransfer.effectAllowed = "none";
            setInvalidDrop(true);
            setTimeout(() => setInvalidDrop(false), 3000);
        }
    };

    return <div
        style={{display: "grid", gridTemplateColumns: "1fr max-content"}}
        onDrop={handleDrop}>
        <input type="file"
               style={{display: "none"}}
               ref={fileInput}
               accept={extensions}
               onChange={handleFileInputChanged}/>
        <input type="text"
               className="theia-input"
               placeholder={placeholder}
               onChange={handleTextInputChanged}
               style={{border: !invalidDrop ? "" : "var(--theia-border-width) solid var(--theia-errorForeground)"}}
               value={file}/>
        <input type="button"
               className="theia-button"
               value="Browse..."
               onClick={() => fileInput.current && fileInput.current.click()}/>
        {children}
    </div>;
};
