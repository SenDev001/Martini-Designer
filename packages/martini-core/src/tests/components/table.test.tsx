import * as Enzyme from "enzyme";
import * as Adapter from "enzyme-adapter-react-16";
import * as React from "react";
import * as Renderer from "react-test-renderer";
import { Table, TableColumn, CellClickEvent } from "../../browser/components/table";

Enzyme.configure({ adapter: new Adapter() });

interface Person {
    firstName: string;
    lastName: string;
    age: number;
}

const columns: TableColumn[] = [
    {
        Header: "First Name",
        accessor: "firstName"
    },
    {
        Header: "Last Name",
        accessor: "lastName"
    },
    {
        Header: "Age",
        accessor: "age"
    }
];

const createPerson = (firstName: string, lastName: string, age: number): Person => ({
    firstName,
    lastName,
    age
});

const people: Person[] = [
    createPerson("Geralt", "Of Rivia", 94),
    createPerson("Yennefer", "Of Vengerberg", 100),
    createPerson("Cirilla", "Of Cintra", 100)
];

test("Table should be rendered with columns and rows", () => {
    const component = Renderer.create(
        <Table columns={columns} data={people} />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test("Table should be rendered with columns but no rows", () => {
    const component = Renderer.create(
        <Table columns={columns} data={[]} />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test("Clicking a cell should call onClick", () => {
    let event: CellClickEvent | undefined;
    const handleCellClick = (e: CellClickEvent) => {
        event = e;
    };
    const table = Enzyme.shallow(
        <Table columns={columns} data={people} onCellClick={handleCellClick} />
    );
    table.find('.td').at(0).simulate("click", {});
    expect(event).not.toBeUndefined();
    expect(event!.columnId).toBe("firstName");
    expect(event!.rowIndex).toBe(0);
});

test("Table should be rendered with the given selected rows", () => {
    const component = Renderer.create(
        <Table
            columns={columns}
            data={people}
            selectedRows={[1, 2]}
        />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test("Pressing arrow down should select the first row", () => {
    let selectedRows: number[] = [];
    const table = Enzyme.shallow(
        <Table
            columns={columns}
            data={people}
            selectedRows={[]}
            onSelectionChange={_selectedRows => selectedRows = _selectedRows}
        />
    );
    table.simulate("keydown", {
        key: "ArrowDown",
        preventDefault: jest.fn()
    });
    expect(selectedRows).toStrictEqual([0]);
});

test("Pressing arrow up should select the last row", () => {
    let selectedRows: number[] = [];
    const table = Enzyme.shallow(
        <Table
            columns={columns}
            data={people}
            selectedRows={[]}
            onSelectionChange={_selectedRows => selectedRows = _selectedRows}
        />
    );
    table.simulate("keydown", {
        key: "ArrowUp",
        preventDefault: jest.fn()
    });
    expect(selectedRows).toStrictEqual([2]);
});

test("Pressing arrow down should select the next row", () => {
    let selectedRows: number[] = [];
    const table = Enzyme.shallow(
        <Table
            columns={columns}
            data={people}
            selectedRows={[0]}
            onSelectionChange={_selectedRows => selectedRows = _selectedRows}
        />
    );
    table.simulate("keydown", {
        key: "ArrowDown",
        preventDefault: jest.fn()
    });
    expect(selectedRows).toStrictEqual([1]);
});

test("Pressing arrow down with last row selected should select the first row", () => {
    let selectedRows: number[] = [];
    const table = Enzyme.shallow(
        <Table
            columns={columns}
            data={people}
            selectedRows={[2]}
            onSelectionChange={_selectedRows => selectedRows = _selectedRows}
        />
    );
    table.simulate("keydown", {
        key: "ArrowDown",
        preventDefault: jest.fn()
    });
    expect(selectedRows).toStrictEqual([0]);
});

test("Pressing arrow up should select the previous row", () => {
    let selectedRows: number[] = [];
    const table = Enzyme.shallow(
        <Table
            columns={columns}
            data={people}
            selectedRows={[1]}
            onSelectionChange={_selectedRows => selectedRows = _selectedRows}
        />
    );
    table.simulate("keydown", {
        key: "ArrowUp",
        preventDefault: jest.fn()
    });
    expect(selectedRows).toStrictEqual([0]);
});

test("Pressing arrow up with the first row selected should select the last row", () => {
    let selectedRows: number[] = [];
    const table = Enzyme.shallow(
        <Table
            columns={columns}
            data={people}
            selectedRows={[0]}
            onSelectionChange={_selectedRows => selectedRows = _selectedRows}
        />
    );
    table.simulate("keydown", {
        key: "ArrowUp",
        preventDefault: jest.fn()
    });
    expect(selectedRows).toStrictEqual([2]);
});

test("Pressing arrow down with shift key should spread the selection to the next row", () => {
    let selectedRows: number[] = [];
    const table = Enzyme.shallow(
        <Table
            columns={columns}
            data={people}
            selectedRows={[0]}
            onSelectionChange={_selectedRows => selectedRows = _selectedRows}
        />
    );
    table.simulate("keydown", {
        key: "ArrowDown",
        shiftKey: true,
        preventDefault: jest.fn()
    });
    expect(selectedRows).toStrictEqual([0, 1]);
});

test("Pressing arrow up with shift key should spread the selection to the previous row", () => {
    let selectedRows: number[] = [];
    const table = Enzyme.shallow(
        <Table
            columns={columns}
            data={people}
            selectedRows={[1]}
            onSelectionChange={_selectedRows => selectedRows = _selectedRows}
        />
    );
    table.simulate("keydown", {
        key: "ArrowUp",
        shiftKey: true,
        preventDefault: jest.fn()
    });
    expect(selectedRows).toStrictEqual([1, 0]);
});

test("Pressing arrow down with shift key should spread the selection to the next row skipping already selected rows", () => {
    let selectedRows: number[] = [];
    const table = Enzyme.shallow(
        <Table
            columns={columns}
            data={[...people, createPerson("", "", 0)]}
            selectedRows={[0, 1, 2]}
            onSelectionChange={_selectedRows => selectedRows = _selectedRows}
        />
    );
    table.simulate("keydown", {
        key: "ArrowDown",
        shiftKey: true,
        preventDefault: jest.fn()
    });
    expect(selectedRows).toStrictEqual([0, 1, 2, 3]);
});

test("Pressing arrow up with shift key should spread the selection to the previous row skipping already selected rows", () => {
    let selectedRows: number[] = [];
    const table = Enzyme.shallow(
        <Table
            columns={columns}
            data={[...people, createPerson("", "", 0)]}
            selectedRows={[3, 2, 1]}
            onSelectionChange={_selectedRows => selectedRows = _selectedRows}
        />
    );
    table.simulate("keydown", {
        key: "ArrowUp",
        shiftKey: true,
        preventDefault: jest.fn()
    });
    expect(selectedRows).toStrictEqual([3, 2, 1, 0]);
});

test("Pressing arrow down should shrink selection", () => {
    let selectedRows: number[] = [];
    const table = Enzyme.shallow(
        <Table
            columns={columns}
            data={[...people]}
            selectedRows={[2, 1, 0]}
            onSelectionChange={_selectedRows => selectedRows = _selectedRows}
        />
    );
    table.simulate("keydown", {
        key: "ArrowDown",
        shiftKey: true,
        preventDefault: jest.fn()
    });
    expect(selectedRows).toStrictEqual([2, 1]);
});

test("Pressing arrow up should shrink selection", () => {
    let selectedRows: number[] = [];
    const table = Enzyme.shallow(
        <Table
            columns={columns}
            data={[...people]}
            selectedRows={[0, 1, 2]}
            onSelectionChange={_selectedRows => selectedRows = _selectedRows}
        />
    );
    table.simulate("keydown", {
        key: "ArrowUp",
        shiftKey: true,
        preventDefault: jest.fn()
    });
    expect(selectedRows).toStrictEqual([0, 1]);
});

test("Clicking a row should select it", () => {
    let selectedRows: number[] = [];
    const table = Enzyme.shallow(
        <Table
            columns={columns}
            data={people}
            selectedRows={[]}
            onSelectionChange={_selectedRows => selectedRows = _selectedRows}
        />
    );
    table.find('.td').at(0).simulate("click", {});
    expect(selectedRows).toStrictEqual([0]);
});

test("Clicking a row should select it and unselect the previous selected row", () => {
    let selectedRows: number[] = [];
    const table = Enzyme.shallow(
        <Table
            columns={columns}
            data={people}
            selectedRows={[0]}
            onSelectionChange={_selectedRows => selectedRows = _selectedRows}
        />
    );
    table.find('.td').at(4).simulate("click", {});
    expect(selectedRows).toStrictEqual([1]);
});

test("Clicking a row with meta key should add it to the selection", () => {
    let selectedRows: number[] = [];
    const table = Enzyme.shallow(
        <Table
            columns={columns}
            data={people}
            selectedRows={[0]}
            onSelectionChange={_selectedRows => selectedRows = _selectedRows}
        />
    );
    table.find('.td').at(4).simulate("click", { metaKey: true });
    expect(selectedRows).toStrictEqual([0, 1]);
});

test("Clicking a selected row with meta key should remove it from the selection", () => {
    let selectedRows: number[] = [];
    const table = Enzyme.shallow(
        <Table
            columns={columns}
            data={people}
            selectedRows={[0, 1]}
            onSelectionChange={_selectedRows => selectedRows = _selectedRows}
        />
    );
    table.find('.td').at(0).simulate("click", { metaKey: true });
    expect(selectedRows).toStrictEqual([1]);
});

test("Clicking a selected row with shift key should select a range of rows", () => {
    let selectedRows: number[] = [];
    const table = Enzyme.shallow(
        <Table
            columns={columns}
            data={people}
            selectedRows={[0]}
            onSelectionChange={_selectedRows => selectedRows = _selectedRows}
        />
    );
    table.find('.td').at(7).simulate("click", {
        shiftKey: true,
        preventDefault: jest.fn()
    });
    expect(selectedRows).toStrictEqual([0, 1, 2]);
});
