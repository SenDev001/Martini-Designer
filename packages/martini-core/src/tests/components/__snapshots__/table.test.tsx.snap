// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should be rendered with columns and rows 1`] = `
<div
  className="sc-AxjAm ljDiNr"
  onKeyDown={[Function]}
>
  <div
    className="table"
    role="table"
    style={
      Object {
        "minWidth": "300px",
      }
    }
  >
    <div
      className="thead"
    >
      <div
        className="tr"
        role="row"
        style={
          Object {
            "display": "flex",
            "flex": "1 0 auto",
            "minWidth": "90px",
          }
        }
      >
        <div
          className="th"
          colSpan={1}
          role="columnheader"
          style={
            Object {
              "alignItems": "center",
              "boxSizing": "border-box",
              "display": "flex",
              "flex": "100 0 auto",
              "justifyContent": "flex-start",
              "minWidth": "30px",
              "overflow": "visible",
              "padding": undefined,
              "position": "relative",
              "width": "100px",
            }
          }
        >
          First Name
          <div
            className="resizer"
            draggable={false}
            onMouseDown={[Function]}
            onTouchStart={[Function]}
            role="separator"
            style={
              Object {
                "cursor": "ew-resize",
              }
            }
          />
        </div>
        <div
          className="th"
          colSpan={1}
          role="columnheader"
          style={
            Object {
              "alignItems": "center",
              "boxSizing": "border-box",
              "display": "flex",
              "flex": "100 0 auto",
              "justifyContent": "flex-start",
              "minWidth": "30px",
              "overflow": "visible",
              "padding": undefined,
              "position": "relative",
              "width": "100px",
            }
          }
        >
          Last Name
          <div
            className="resizer"
            draggable={false}
            onMouseDown={[Function]}
            onTouchStart={[Function]}
            role="separator"
            style={
              Object {
                "cursor": "ew-resize",
              }
            }
          />
        </div>
        <div
          className="th"
          colSpan={1}
          role="columnheader"
          style={
            Object {
              "alignItems": "center",
              "boxSizing": "border-box",
              "display": "flex",
              "flex": "100 0 auto",
              "justifyContent": "flex-start",
              "minWidth": "30px",
              "overflow": "visible",
              "padding": undefined,
              "position": "relative",
              "width": "100px",
            }
          }
        >
          Age
          <div
            className="resizer"
            draggable={false}
            onMouseDown={[Function]}
            onTouchStart={[Function]}
            role="separator"
            style={
              Object {
                "cursor": "ew-resize",
              }
            }
          />
        </div>
      </div>
    </div>
    <div
      className="tbody"
    >
      <div
        className="tr"
        role="row"
        style={
          Object {
            "display": "flex",
            "flex": "1 0 auto",
            "minWidth": "90px",
          }
        }
      >
        <div
          className="td"
          onClick={[Function]}
          onContextMenu={[Function]}
          onDoubleClick={[Function]}
          role="cell"
          style={
            Object {
              "alignItems": "center",
              "boxSizing": "border-box",
              "display": "flex",
              "flex": "100 0 auto",
              "justifyContent": "flex-start",
              "minWidth": "30px",
              "overflow": "visible",
              "padding": undefined,
              "width": "100px",
            }
          }
        >
          Geralt
        </div>
        <div
          className="td"
          onClick={[Function]}
          onContextMenu={[Function]}
          onDoubleClick={[Function]}
          role="cell"
          style={
            Object {
              "alignItems": "center",
              "boxSizing": "border-box",
              "display": "flex",
              "flex": "100 0 auto",
              "justifyContent": "flex-start",
              "minWidth": "30px",
              "overflow": "visible",
              "padding": undefined,
              "width": "100px",
            }
          }
        >
          Of Rivia
        </div>
        <div
          className="td"
          onClick={[Function]}
          onContextMenu={[Function]}
          onDoubleClick={[Function]}
          role="cell"
          style={
            Object {
              "alignItems": "center",
              "boxSizing": "border-box",
              "display": "flex",
              "flex": "100 0 auto",
              "justifyContent": "flex-start",
              "minWidth": "30px",
              "overflow": "visible",
              "padding": undefined,
              "width": "100px",
            }
          }
        >
          94
        </div>
      </div>
      <div
        className="tr"
        role="row"
        style={
          Object {
            "display": "flex",
            "flex": "1 0 auto",
            "minWidth": "90px",
          }
        }
      >
        <div
          className="td"
          onClick={[Function]}
          onContextMenu={[Function]}
          onDoubleClick={[Function]}
          role="cell"
          style={
            Object {
              "alignItems": "center",
              "boxSizing": "border-box",
              "display": "flex",
              "flex": "100 0 auto",
              "justifyContent": "flex-start",
              "minWidth": "30px",
              "overflow": "visible",
              "padding": undefined,
              "width": "100px",
            }
          }
        >
          Yennefer
        </div>
        <div
          className="td"
          onClick={[Function]}
          onContextMenu={[Function]}
          onDoubleClick={[Function]}
          role="cell"
          style={
            Object {
              "alignItems": "center",
              "boxSizing": "border-box",
              "display": "flex",
              "flex": "100 0 auto",
              "justifyContent": "flex-start",
              "minWidth": "30px",
              "overflow": "visible",
              "padding": undefined,
              "width": "100px",
            }
          }
        >
          Of Vengerberg
        </div>
        <div
          className="td"
          onClick={[Function]}
          onContextMenu={[Function]}
          onDoubleClick={[Function]}
          role="cell"
          style={
            Object {
              "alignItems": "center",
              "boxSizing": "border-box",
              "display": "flex",
              "flex": "100 0 auto",
              "justifyContent": "flex-start",
              "minWidth": "30px",
              "overflow": "visible",
              "padding": undefined,
              "width": "100px",
            }
          }
        >
          100
        </div>
      </div>
      <div
        className="tr"
        role="row"
        style={
          Object {
            "display": "flex",
            "flex": "1 0 auto",
            "minWidth": "90px",
          }
        }
      >
        <div
          className="td"
          onClick={[Function]}
          onContextMenu={[Function]}
          onDoubleClick={[Function]}
          role="cell"
          style={
            Object {
              "alignItems": "center",
              "boxSizing": "border-box",
              "display": "flex",
              "flex": "100 0 auto",
              "justifyContent": "flex-start",
              "minWidth": "30px",
              "overflow": "visible",
              "padding": undefined,
              "width": "100px",
            }
          }
        >
          Cirilla
        </div>
        <div
          className="td"
          onClick={[Function]}
          onContextMenu={[Function]}
          onDoubleClick={[Function]}
          role="cell"
          style={
            Object {
              "alignItems": "center",
              "boxSizing": "border-box",
              "display": "flex",
              "flex": "100 0 auto",
              "justifyContent": "flex-start",
              "minWidth": "30px",
              "overflow": "visible",
              "padding": undefined,
              "width": "100px",
            }
          }
        >
          Of Cintra
        </div>
        <div
          className="td"
          onClick={[Function]}
          onContextMenu={[Function]}
          onDoubleClick={[Function]}
          role="cell"
          style={
            Object {
              "alignItems": "center",
              "boxSizing": "border-box",
              "display": "flex",
              "flex": "100 0 auto",
              "justifyContent": "flex-start",
              "minWidth": "30px",
              "overflow": "visible",
              "padding": undefined,
              "width": "100px",
            }
          }
        >
          100
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Table should be rendered with columns but no rows 1`] = `
<div
  className="sc-AxjAm ljDiNr"
  onKeyDown={[Function]}
>
  <div
    className="table"
    role="table"
    style={
      Object {
        "minWidth": "300px",
      }
    }
  >
    <div
      className="thead"
    >
      <div
        className="tr"
        role="row"
        style={
          Object {
            "display": "flex",
            "flex": "1 0 auto",
            "minWidth": "90px",
          }
        }
      >
        <div
          className="th"
          colSpan={1}
          role="columnheader"
          style={
            Object {
              "alignItems": "center",
              "boxSizing": "border-box",
              "display": "flex",
              "flex": "100 0 auto",
              "justifyContent": "flex-start",
              "minWidth": "30px",
              "overflow": "visible",
              "padding": undefined,
              "position": "relative",
              "width": "100px",
            }
          }
        >
          First Name
          <div
            className="resizer"
            draggable={false}
            onMouseDown={[Function]}
            onTouchStart={[Function]}
            role="separator"
            style={
              Object {
                "cursor": "ew-resize",
              }
            }
          />
        </div>
        <div
          className="th"
          colSpan={1}
          role="columnheader"
          style={
            Object {
              "alignItems": "center",
              "boxSizing": "border-box",
              "display": "flex",
              "flex": "100 0 auto",
              "justifyContent": "flex-start",
              "minWidth": "30px",
              "overflow": "visible",
              "padding": undefined,
              "position": "relative",
              "width": "100px",
            }
          }
        >
          Last Name
          <div
            className="resizer"
            draggable={false}
            onMouseDown={[Function]}
            onTouchStart={[Function]}
            role="separator"
            style={
              Object {
                "cursor": "ew-resize",
              }
            }
          />
        </div>
        <div
          className="th"
          colSpan={1}
          role="columnheader"
          style={
            Object {
              "alignItems": "center",
              "boxSizing": "border-box",
              "display": "flex",
              "flex": "100 0 auto",
              "justifyContent": "flex-start",
              "minWidth": "30px",
              "overflow": "visible",
              "padding": undefined,
              "position": "relative",
              "width": "100px",
            }
          }
        >
          Age
          <div
            className="resizer"
            draggable={false}
            onMouseDown={[Function]}
            onTouchStart={[Function]}
            role="separator"
            style={
              Object {
                "cursor": "ew-resize",
              }
            }
          />
        </div>
      </div>
    </div>
    <div
      className="tbody"
    />
  </div>
</div>
`;

exports[`Table should be rendered with the given selected rows 1`] = `
<div
  className="sc-AxjAm ljDiNr"
  onKeyDown={[Function]}
>
  <div
    className="table"
    role="table"
    style={
      Object {
        "minWidth": "300px",
      }
    }
  >
    <div
      className="thead"
    >
      <div
        className="tr"
        role="row"
        style={
          Object {
            "display": "flex",
            "flex": "1 0 auto",
            "minWidth": "90px",
          }
        }
      >
        <div
          className="th"
          colSpan={1}
          role="columnheader"
          style={
            Object {
              "alignItems": "center",
              "boxSizing": "border-box",
              "display": "flex",
              "flex": "100 0 auto",
              "justifyContent": "flex-start",
              "minWidth": "30px",
              "overflow": "visible",
              "padding": undefined,
              "position": "relative",
              "width": "100px",
            }
          }
        >
          First Name
          <div
            className="resizer"
            draggable={false}
            onMouseDown={[Function]}
            onTouchStart={[Function]}
            role="separator"
            style={
              Object {
                "cursor": "ew-resize",
              }
            }
          />
        </div>
        <div
          className="th"
          colSpan={1}
          role="columnheader"
          style={
            Object {
              "alignItems": "center",
              "boxSizing": "border-box",
              "display": "flex",
              "flex": "100 0 auto",
              "justifyContent": "flex-start",
              "minWidth": "30px",
              "overflow": "visible",
              "padding": undefined,
              "position": "relative",
              "width": "100px",
            }
          }
        >
          Last Name
          <div
            className="resizer"
            draggable={false}
            onMouseDown={[Function]}
            onTouchStart={[Function]}
            role="separator"
            style={
              Object {
                "cursor": "ew-resize",
              }
            }
          />
        </div>
        <div
          className="th"
          colSpan={1}
          role="columnheader"
          style={
            Object {
              "alignItems": "center",
              "boxSizing": "border-box",
              "display": "flex",
              "flex": "100 0 auto",
              "justifyContent": "flex-start",
              "minWidth": "30px",
              "overflow": "visible",
              "padding": undefined,
              "position": "relative",
              "width": "100px",
            }
          }
        >
          Age
          <div
            className="resizer"
            draggable={false}
            onMouseDown={[Function]}
            onTouchStart={[Function]}
            role="separator"
            style={
              Object {
                "cursor": "ew-resize",
              }
            }
          />
        </div>
      </div>
    </div>
    <div
      className="tbody"
    >
      <div
        className="tr"
        role="row"
        style={
          Object {
            "display": "flex",
            "flex": "1 0 auto",
            "minWidth": "90px",
          }
        }
      >
        <div
          className="td"
          onClick={[Function]}
          onContextMenu={[Function]}
          onDoubleClick={[Function]}
          role="cell"
          style={
            Object {
              "alignItems": "center",
              "boxSizing": "border-box",
              "display": "flex",
              "flex": "100 0 auto",
              "justifyContent": "flex-start",
              "minWidth": "30px",
              "overflow": "visible",
              "padding": undefined,
              "width": "100px",
            }
          }
        >
          Geralt
        </div>
        <div
          className="td"
          onClick={[Function]}
          onContextMenu={[Function]}
          onDoubleClick={[Function]}
          role="cell"
          style={
            Object {
              "alignItems": "center",
              "boxSizing": "border-box",
              "display": "flex",
              "flex": "100 0 auto",
              "justifyContent": "flex-start",
              "minWidth": "30px",
              "overflow": "visible",
              "padding": undefined,
              "width": "100px",
            }
          }
        >
          Of Rivia
        </div>
        <div
          className="td"
          onClick={[Function]}
          onContextMenu={[Function]}
          onDoubleClick={[Function]}
          role="cell"
          style={
            Object {
              "alignItems": "center",
              "boxSizing": "border-box",
              "display": "flex",
              "flex": "100 0 auto",
              "justifyContent": "flex-start",
              "minWidth": "30px",
              "overflow": "visible",
              "padding": undefined,
              "width": "100px",
            }
          }
        >
          94
        </div>
      </div>
      <div
        className="tr"
        role="row"
        style={
          Object {
            "display": "flex",
            "flex": "1 0 auto",
            "minWidth": "90px",
          }
        }
      >
        <div
          className="td selected"
          onClick={[Function]}
          onContextMenu={[Function]}
          onDoubleClick={[Function]}
          role="cell"
          style={
            Object {
              "alignItems": "center",
              "boxSizing": "border-box",
              "display": "flex",
              "flex": "100 0 auto",
              "justifyContent": "flex-start",
              "minWidth": "30px",
              "overflow": "visible",
              "padding": undefined,
              "width": "100px",
            }
          }
        >
          Yennefer
        </div>
        <div
          className="td selected"
          onClick={[Function]}
          onContextMenu={[Function]}
          onDoubleClick={[Function]}
          role="cell"
          style={
            Object {
              "alignItems": "center",
              "boxSizing": "border-box",
              "display": "flex",
              "flex": "100 0 auto",
              "justifyContent": "flex-start",
              "minWidth": "30px",
              "overflow": "visible",
              "padding": undefined,
              "width": "100px",
            }
          }
        >
          Of Vengerberg
        </div>
        <div
          className="td selected"
          onClick={[Function]}
          onContextMenu={[Function]}
          onDoubleClick={[Function]}
          role="cell"
          style={
            Object {
              "alignItems": "center",
              "boxSizing": "border-box",
              "display": "flex",
              "flex": "100 0 auto",
              "justifyContent": "flex-start",
              "minWidth": "30px",
              "overflow": "visible",
              "padding": undefined,
              "width": "100px",
            }
          }
        >
          100
        </div>
      </div>
      <div
        className="tr"
        role="row"
        style={
          Object {
            "display": "flex",
            "flex": "1 0 auto",
            "minWidth": "90px",
          }
        }
      >
        <div
          className="td selected"
          onClick={[Function]}
          onContextMenu={[Function]}
          onDoubleClick={[Function]}
          role="cell"
          style={
            Object {
              "alignItems": "center",
              "boxSizing": "border-box",
              "display": "flex",
              "flex": "100 0 auto",
              "justifyContent": "flex-start",
              "minWidth": "30px",
              "overflow": "visible",
              "padding": undefined,
              "width": "100px",
            }
          }
        >
          Cirilla
        </div>
        <div
          className="td selected"
          onClick={[Function]}
          onContextMenu={[Function]}
          onDoubleClick={[Function]}
          role="cell"
          style={
            Object {
              "alignItems": "center",
              "boxSizing": "border-box",
              "display": "flex",
              "flex": "100 0 auto",
              "justifyContent": "flex-start",
              "minWidth": "30px",
              "overflow": "visible",
              "padding": undefined,
              "width": "100px",
            }
          }
        >
          Of Cintra
        </div>
        <div
          className="td selected"
          onClick={[Function]}
          onContextMenu={[Function]}
          onDoubleClick={[Function]}
          role="cell"
          style={
            Object {
              "alignItems": "center",
              "boxSizing": "border-box",
              "display": "flex",
              "flex": "100 0 auto",
              "justifyContent": "flex-start",
              "minWidth": "30px",
              "overflow": "visible",
              "padding": undefined,
              "width": "100px",
            }
          }
        >
          100
        </div>
      </div>
    </div>
  </div>
</div>
`;
